src/Processors/Executors/PipelineExecutor.cpp

```diff
@@ -399,7 +399,18 @@ void PipelineExecutor::executeImpl(size_t num_threads, bool concurrency_control)
 
     if (num_threads > 1)
     {
-        spawnThreads(); // start at least one thread
+        try
+        {
+            spawnThreads(); // start at least one thread
+        }
+        catch (...)
+        {
+            /// spawnThreads can throw an exception, for example CANNOT_SCHEDULE_TASK.
+            /// We should cancel execution properly before rethrow.
+            cancel();
```

Comment:
Shouldn't we rather wrap the entire `executeImpl()` body in try/catch and call `cancel()` accordingly? I find this fix too narrow

Reply:
But `spawnThreads` is the only place where we don't cancel execution in case of an exception. Other code either doesn't throw or handles exceptions correctly. Ideally, `spawnThreads` should not throw, but by now it's not possible: https://github.com/ClickHouse/clickhouse-core-incidents/issues/10#issuecomment-1816103879

Reply:
No pressure regarding merging the fix - feel free to, but ...
> But spawnThreads is the only place ...

It seems, it's a strong statement which can change overtime. But the function should hold the contract, i.e. - in case of exception, cancel whatever background jobs/processes it's doing. If it'd be implemented in the proposed way, then we'd have the escalation in the first place.

Reply:
Actually, I just realized that there is a code above:
```
    SCOPE_EXIT_SAFE(
        if (!finished_flag)
        {
            finish();
            if (pool)
                pool->wait();
        }
    );
```

And AFAIU `finished_flag` can be `false` only in case of an exception in `executeImpl()`, but we should call `cancel()` instead of `finish()` in this case, it's similar to https://github.com/ClickHouse/ClickHouse/pull/52533. So, let's do it